python/ml4ir/base/model/architectures/dnn.py

```diff
@@ -11,12 +17,30 @@ class DNNLayer:
 
 
 class DNN:
-    def __init__(self, model_config):
-        self.layer_ops: List = self.define_architecture(model_config)
+    def __init__(self, model_config: dict, feature_config: dict, file_io):
+        self.file_io: FileIO = file_io
+        self.layer_ops: List = self.define_architecture(model_config, feature_config)
+
+    def define_architecture(self, model_config: dict, feature_config: dict):
+        """
+        Convert the model from model_config to a List of tensorflow.keras.layer
 
-    def define_architecture(self, model_config):
+        :param model_config: dict corresponding to the model config
+        :param feature_config: dict corresponding to the feature config, only used in case of classification if the last
+            layer of the model_config doesn't have a units number defined (or set to -1). In which case we retrieve the
+            label vocabulary defined in the feature_config to deduce the number of units.
+        :return: List[layers]: list of keras layer corresponding to each of the layers defined in the model_config.
+        """
         def get_op(layer_type, layer_args):
             if layer_type == DNNLayer.DENSE:
+                if not "units" in layer_args or layer_args["units"] == -1:
+                    try:
+                        label_feature_info = feature_config.get_label()
+                        vocabulary_keys, vocabulary_ids = get_vocabulary_info(label_feature_info, self.file_io)
+                        layer_args["units"] = len(vocabulary_keys) + OOV
```

Comment:
This is nice!

Reply:
Although this looks very neat and is handy for the k-class classification problem, I am not a fan of wiring in feature config and vocabulary here as it does not make sense for other ML applications. If I want to train k-class classifier without a vocabulary file, how would I do it? I think the vocabulary logic should be outside of this architecture function.

Reply:
If I want to train k-class classifier without a vocabulary file, how would I do it? ==> just add the "units" param in the model_config.

Where the confusion may come from is the name "feature config", which isn't actually only "feature config" in the sense of input, but it also contains the labels/output definition.

Reply:
@Ullimague and I synced offline and decided that this is okay for now and we will plan a redesign of feature config and model config in the future.